extensions/dialog/internal.m

```diff
@@ -0,0 +1,388 @@
+///  * [style] - An optional style of the dialog box as a string. Defaults to "NSWarningAlertStyle".
+///
+/// Returns:
+///  * The value of the button as a string.
+///
+/// Notes:
+///  * This alert is blocking (i.e. no other Lua code will be processed until the alert is closed).
+///  * The optional values must be entered in order (i.e. you can't supply `style` without also supplying `buttonOne` and `buttonTwo`).
+///  * [style] can be "NSWarningAlertStyle", "NSInformationalAlertStyle" or "NSCriticalAlertStyle". If something other than these string values is given, it will use "NSWarningAlertStyle".
+/*
+ EXAMPLE:
+ test = hs.webview.newBrowser(hs.geometry.rect(250, 250, 250, 250)):show()
+ hs.dialog.webviewAlert(test, function(result) print("Callback Result: " .. result) end, "Message", "Informative Text", "Button One", "Button Two", "NSCriticalAlertStyle")
+ */
+static int webviewAlert(lua_State *L) {
+    
+    NSString* defaultButton = @"OK";
+    
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, "hs.webview", LS_TFUNCTION, LS_TSTRING, LS_TSTRING, LS_TSTRING | LS_TOPTIONAL, LS_TSTRING | LS_TOPTIONAL, LS_TSTRING | LS_TOPTIONAL, LS_TBREAK];
+    
+    NSWindow *webview = [skin toNSObjectAtIndex:1];
+    //NSObject *callback = [skin toNSObjectAtIndex:2];
+    NSString *message = [skin toNSObjectAtIndex:3];
+    NSString *informativeText = [skin toNSObjectAtIndex:4];
+    NSString *buttonOne = [skin toNSObjectAtIndex:5];
+    NSString *buttonTwo = [skin toNSObjectAtIndex:6];
+    NSString *style = [skin toNSObjectAtIndex:7];
+    
+    NSAlert *alert = [[NSAlert alloc] init];
+    [alert setMessageText:message];
+    [alert setInformativeText:informativeText];
+    
+    if( buttonOne == nil ){
+        [alert addButtonWithTitle:defaultButton];
+    }
+    else
+    {
+        [alert addButtonWithTitle:buttonOne];
+    }
+    
+    if (buttonTwo != nil ) {
+        [alert addButtonWithTitle:buttonTwo];
+    }
+    
+    if (style == nil){
+        [alert setAlertStyle:NSWarningAlertStyle];
+    }
+    else
+    {
+        if ([style isEqualToString:@"NSWarningAlertStyle"]) {
+            [alert setAlertStyle:NSWarningAlertStyle];
+        }
+        else if ([style isEqualToString:@"NSInformationalAlertStyle"]) {
+            [alert setAlertStyle:NSInformationalAlertStyle];
+        }
+        else if ([style isEqualToString:@"NSCriticalAlertStyle"]) {
+            [alert setAlertStyle:NSCriticalAlertStyle];
+        }
+        else
+        {
+            [alert setAlertStyle:NSWarningAlertStyle];
+        }
+    }
+    
+    [alert beginSheetModalForWindow:webview completionHandler:^(NSModalResponse result){
+        if (result == NSAlertFirstButtonReturn) {
+            if (buttonOne == nil) {
+                [LuaSkin logInfo:@"hs.dialog.webviewAlert() - Default button pressed during callback."];
+                
+                // @asmagill - what do I need to put here to trigger the callback?
```

Comment:
Something along the lines of

~~~objc
    lua_pushvalue(L, 2) ; // push the function passed in as an argument to the top of the stack
    [skin pushNSObject:defaultButton]; // or whatever arguments you want passed to the callback
    if (![skin protectedCallAndTraceback:1 nresults:0]) { // returns NO on error, so we check if the result is !YES
        [skin logError:[NSString stringWithFormat:@"hs.dialog:callback error - %s", lua_tostring(L, -1)] ; // -1 indicates the top item of the stack, which will be an error message string in this case
        lua_pop(L, 1) ; // remove the error from the stack to keep it clean
    }
~~~

should probably work.

Since you're calling the same callback with basically the same format, just a different argument, I'd probably use the if-then-else-etc to set an NSString containing the name of the button (i.e. defaultButton or buttonOne or etc.) and then only do the callback code once with that variable after the if-then-else block, just to reduce repetition and chance of typos.

Reply:
The only thing to watch for is I don't know what thread alert will run its code it... I *think* the main one, but if you experience some unexpected crashes, check the Console and User crashlog... if the crash is in any thread but thread 0, wrap the whole thing in 

~~~objc
dispatch_async(dispatch_get_main_queue(), ^{
    ...  the above code for doing the callback ...
}) ;
~~~

Reply:
@asmagill - I tried using `lua_pushvalue(L, 2)`, but it just gives me: 

```
2017-08-30 14:29:15: ********
2017-08-30 14:29:15: 14:29:15 ERROR:   LuaSkin: hs.dialog:callback error - attempt to call a nil value
stack traceback:
2017-08-30 14:29:15: ********
```

I've updated the branch to reflect the changed code. Any ideas? I'm probably just doing something silly.

Although I haven't seen any crashes, I did try wrapping everything in `dispatch_async` - but it gave me some Xcode error, so I'll just leave it for now.